activerecord/test/cases/finder_test.rb

```diff
@@ -265,6 +265,11 @@ def test_find_by_sql_with_sti_on_joined_table
     assert_equal [Account], accounts.collect(&:class).uniq
   end
 
+  def test_find_by_association_subquery
+    author = authors(:david)
+    assert_equal author.post, Post.find_by(author_id: Author.where(id: author))
```

Comment:
I added a test.

> Btw, is that case working right now, or this patch fix that too?

That case works without this patch because of the following code.

https://github.com/rails/rails/blob/master/activerecord/lib/active_record/core.rb#L184
```ruby
# core.rb line 184
return super unless hash.keys.all? { |k| columns_hash.has_key?(k.to_s) }
```